x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java

```diff
@@ -345,18 +364,25 @@ void extractToken(Consumer<AuthenticationToken> consumer) {
             action.run();
         }
 
+        private void consumeToken(AuthenticationToken token){
+            consumeToken(token, null);
+        }
+
         /**
          * Consumes the {@link AuthenticationToken} provided by the caller. In the case of a {@code null} token, {@link #handleNullToken()}
-         * is called. In the case of a {@code non-null} token, the realms are iterated over and the first realm that returns a non-null
-         * {@link User} is the authenticating realm and iteration is stopped. This user is then passed to {@link #consumeUser(User, Map)}
-         * if no exception was caught while trying to authenticate the token
+         * is called. In the case of a {@code non-null} token, the realms are iterated over in the order defined in the configuration
+         * while possible also taking into consideration the last realm that authenticated this principal. If {@code preferredRealm} is
+         * not null, only this realm will be attempted, regardless of its order/applicability. When consulting multiple realms, the
+         * first realm that returns a non-null {@link User} is the authenticating realm and iteration is stopped. This user is then
+         * passed to {@link #consumeUser(User, Map)} if no exception was caught while trying to authenticate the token
          */
-        private void consumeToken(AuthenticationToken token) {
+        private void consumeToken(AuthenticationToken token, @Nullable RealmConfig.RealmIdentifier preferredRealm) {
```

Comment:
I think the `AuthenticationService` is too generic to handle the particular logic of the OIDC and SAML realms.

This method's core logic is to iterate the realms list and try to authenticate the token. I don't like it that we upset this logic in this case. I suggest we either create a new method in the `AuthenticationService` which only deals with a single realm, or more preferable for me, would be to incorporate the realm "preference" (which is more than a preference) inside the `AuthenticationToken`.

For example, in `OpenIdConnectRealm`:
```
    public boolean supports(AuthenticationToken token) {
        If (false == (token instanceof OpenIdConnectToken)) {
            return false;
        }
        (OpenIdConnectToken) oidcToken = (OpenIdConnectToken) token;
        if (null == oidcToken.getRealmName()) {
            return true;
        } else {
            return oidcToken.getRealmName().equals(this.name());
        }
    }
```
This alternative moves the realm "preference" logic to the individual realm, away from the more generic `AuthenticationService`. This allows, for example, to have `OIDC authenticate` API calls specify an `iss` or a `realm` the same way that `OIDC prepareAuthenticate` can specify an `iss` or a `realm`. You can see how the `iss` request parameter, which is not relevant for the SAML realm, is handled internally by the OIDC realm.

What do you think?

Reply:
Thanks for the feedback @albertzaharovits , appreciated as always!

>  I don't like it that we upset this logic in this case. 

Could you add some context and/or arguments describing what you don't like please ?

> I suggest we either create a new method in the AuthenticationService which only deals with a single realm

I think that adding a different `consumeToken` method which will share much of the logic of the existing one doesn't buy us anything. I believe this fits nicely here, the same way the reordering of the realm list based on the cache fits. 

> or more preferable for me, would be to incorporate the realm "preference" (which is more than a preference) inside the AuthenticationToken.

I considered dealing with this in the AuthenticationToken but since we want to remove the AuthenticationToken concept entirely, it didn't seem like a good idea to stuff more things in there. Also, why replicate the exact same logic in both realms ? 

> This allows, for example, to have OIDC authenticate API calls specify an iss or a realm the same way that OIDC prepareAuthenticate can specify an iss or a realm. You can see how the iss request parameter, which is not relevant for the SAML realm, is handled internally by the OIDC realm.

There is no use passing `iss` or any other parameter. The use case this change tries to solve is similar for all our web SSO realms ( OIDC and SAML for the time being ) and it is that we know which realm needs to consume the AuthenticationToken so we don't need to iterate through the rest. 

>  "preference" (which is more than a preference)

In retrospect, `preferredRealm` wasn't a good choice for the name, but everything else above still stands. 

I'm open to being swayed with more arguments if you feel strongly about this

Reply:
> Could you add some context and/or arguments describing what you don't like please ?

The `AuthenticationService`in the `consumeToken` method iterates the realms list, checks if each supports the token and tries to authenticate (SUCCESS, CONTINUE, TERMINATE). Crucially, the realms list is private; the caller (to `AuthenticationService#authenticate`) has no option to reorder or truncate it. For this reason the caller isn't even concerned about what a realm is. The proposed change breaks this encapsulation, because the caller can now "specify a preference about a realm".

On a more pragmatic, lower level, the proposed changes make the code more complex. It probably works, but it is more complex. The method builds the realms list as well as iterates over it (in a very particular way). This can be split into two. For example, one method receives the realms list as a parameter, and there is a different method that builds the realms list and calls the other method.

But, aside from breaking the encapsulation, is it really safe for the caller to alter the realms list? For example, a successful OIDC token authentication changes the realm order for all successive OIDC tokens. Why should a successful authentication when the caller specifies a realm preference affect the order for OIDC token authentications when the caller does NOT specify an order? This is a rhetorical question. I want to point out that whatever assumptions we had mentally when inspecting the code when the realms list was hidden, do not hold anymore when the caller has control on the list. For that matter, only slightly exaggerating for emphasis, I would prefer to just pull the specific OIDC realm by name and call `authenticate` on it directly, bypassing the `AuthenticationService` altogether.

> I considered dealing with this in the AuthenticationToken but since we want to remove the AuthenticationToken concept entirely, it didn't seem like a good idea to stuff more things in there.

I don't think this is a good idea. The code that is added should fit with what exists right now. How can we know that in the future, when the `AuthenticationToken` is removed, the proposed code change in the `AuthenticationService` is still valid? Since the `AuthenticationService` generates and consumes `AuthenticationToken`s, it is very probable that it will not be.

> Also, why replicate the exact same logic in both realms ?

I think the replicated code (`token.getRealm().equals(this.name())`) is small enough so that this is not a problem.

> There is no use passing iss or any other parameter. The use case this change tries to solve is similar for all our web SSO realms ( OIDC and SAML for the time being ) and it is that we know which realm needs to consume the AuthenticationToken so we don't need to iterate through the rest.

I _think_ I understand this. Though this is not strictly related to the matter being discussed, I think it has an impact on the implementation so I'll ask: The client can call `OIDC prepare` and `OIDC authenticated` using the `realm` request parameter (for both calls). Is it useful for it to be able to call the same APIs using the `iss` request parameter (for both calls)? If not, why would the client use `iss` over the `realm` parameter for the `OIDC prepare` call?

Reply:
>  Crucially, the realms list is private; the caller (to AuthenticationService#authenticate) has no option to reorder or truncate it. For this reason the caller isn't even concerned about what a realm is. The proposed change breaks this encapsulation, because the caller can now "specify a preference about a realm".

Previously, the caller wasn't actually concerned about the realm. Now the caller is a caller that is aware of what a realm is and which one should be used. I see what you're saying but I think that breaking this encapsulation by _only_ filtering 

> On a more pragmatic, lower level, the proposed changes make the code more complex. It probably works, but it is more complex. The method builds the realms list as well as iterates over it (in a very particular way).

Didn't it do that before the proposed changes also? 

> This can be split into two. For example, one method receives the realms list as a parameter, and there is a different method that builds the realms list and calls the other method.

Not really opposed to breaking this up but filtering first, gains us to not have to hit the cache and reorder the list if we don't have to. 

>  Why should a successful authentication when the caller specifies a realm preference affect the order for OIDC token authentications when the caller does NOT specify an order?

Not sure where you're getting this from. Why would this happen ? Wouldn't subsequent OIDC token authentications still call `getRealmList()` with a `Null` `preferredRealm` and get the behavior they get today? What am I missing here? 

> I don't think this is a good idea. The code that is added should fit with what exists right now. How can we know that in the future

I would agree with you if the AuthenticationToken solution was the obvious best solution, but I don't believe it is. This gets a little too theoretical for the sake of argumentation, but I have always seen us being forward looking when it comes to changes we know we want to do and that this was a valid argument

> I think the replicated code (token.getRealm().equals(this.name())) is small enough so that this is not a problem.

Fair enough

> I think I understand this. Though this is not strictly related to the matter being discussed, I think it has an impact on the implementation so I'll ask: The client can call OIDC prepare and OIDC authenticated using the realm request parameter (for both calls). Is it useful for it to be able to call the same APIs using the iss request parameter (for both calls)? If not, why would the client use iss over the realm parameter for the OIDC prepare call?

The `iss` is a parameter in the OIDC `prepare` endpoint as it is used to handle [3rd party initiated login](https://openid.net/specs/openid-connect-core-1_0.html#ThirdPartyInitiatedLogin). It is used there since the initiator of the login knows what the OPs Issuer is but doesn't know what we call the Elasticsearch realm (or what an Elasticsearch realm is )
`iss` _can't_ be used in the `authenticate` endpoint as it has no meaning there, the OP needs to be sending the code that we will exchange for an ID Token ( authorization code grant ) or the ID Token itself ( implicit flow ). We can't do anything if it sends the issuer name again to us.  

see: 

<img src="https://docs.google.com/drawings/d/e/2PACX-1vRLBzv8ApWxylFDkvMcDsCsNuhBraAeP_MQa5IbUnFxF-taeSznajgMmZvybya2xA7FL567WTPztMWp/pub?w=1504&amp;h=1145">


I don't feel like we should put much more time/energy in dueling over one. I'll go ahead with a solution that doesn't break this encapsulation and ping for a review